sirepo/template/template_common.py

```diff
@@ -119,6 +120,170 @@ def __scale_model(self, name, model, is_native):
+                    version='3.7.2'
+                )
+            )
+        )
+
+    @classmethod
+    def _dict_to_kwargs_str(cls, d):
+        d_str = ''
+        for k in d:
+            v = f'\'{d[k]}\'' if isinstance(d[k], pykern.pkconfig.STRING_TYPES) \
+                else f'{d[k]}'
+            d_str += f'{k}={v},'
+        return d_str
+
+    def __init__(self, sim_type, data):
+        self.data = data
+        self.notebook = JupyterNotebook._base_dict()
+        self.imports = PKDict()
+        self.widgets = []
+
+        # cell 0
+        self.add_markdown_cell(
+            [
+                '# {} - {}'.format(sim_type, data.models.simulation.name),
+            ]
+        )
+        # cell 1
+        self.add_markdown_cell(
+            ['## Imports',]
+        )
+        # cell 2
+        self.add_code_cell([])
+
+        super(object, self).__init__()
+
+    def add_cell(self, cell_type, source_strings, hide=False):
+        assert cell_type in self._CELL_TYPES, 'Invalid cell type {}'.format(cell_type)
+        cell = PKDict(
+            cell_type=cell_type,
+            metadata={},
+            source=[s + ('\n' if s[-1] != '\n' else '') for s in source_strings]
+        )
+        if hide:
+            cell.metadata['jupyter'] = {
+                'source_hidden': 'true'
+            }
+        self.notebook.cells.append(cell)
+
+    def add_code_cell(self, source_strings, hide=False):
+        self.add_cell('code', source_strings, hide=hide)
+
+    # {<pkg>: [sub_pkg]}
+    # just merge?
+    def add_imports(self, pkg_dict):
+        for pkg in pkg_dict:
+            if pkg not in self.imports:
+                self.imports[pkg] = pkg_dict[pkg]
+        for s in pkg_dict[pkg]:
+            if s not in self.imports[pkg]:
+                self.imports[pkg].append(s)
+        self._update()
+
+    # meaningful to load arbitrary file name?
+    def add_load_csv(self, widget_var):
+        data_var = f'data_{widget_var}'
+        self.add_imports({'numpy': ['genfromtxt'], })
+        self.add_code_cell(
+            [
+                f'f = {widget_var}.value',
+                'f_name = next(iter(f.keys()))',
+                f'{data_var} = numpy.genfromtxt(f_name, delimiter=\',\')',
+            ]
+        )
+        return data_var
+
+    def add_markdown_cell(self, source_strings):
+        self.add_cell('markdown', source_strings)
+
+    # parameter plot
+    def add_report(self, cfg):
+        self.add_imports({'matplotlib': ['pyplot']})
+        plot_strs = []
+        legends = []
+        for y_cfg in cfg.y_info:
+            x_pts_var = y_cfg.x_points if 'x_points' in y_cfg else cfg.x_var
+            plot_strs.append(f'pyplot.plot({x_pts_var}, {y_cfg.y_var}, \'{self._PYPLOT_STYLE_MAP[y_cfg.style]}\')')
+            legends.append(f'\'{y_cfg.y_label}\'')
+        code = [
+                'pyplot.figure()',
+                f'pyplot.xlabel(\'{cfg.x_label}\')',
+                f'pyplot.legend({legends})',
```

Comment:
using an f-string here is a bit odd, because it implies that f-string will convert a list into proper python form, whereas I think f-string is always human readable (str). You want to say f'{legends!r}' so that it ensures that the object is python parseable.

Reply:
2ea5753